sinbin/src/Pos/Sinbin/Ssc/VssCertData.hs

```diff
@@ -0,0 +1,265 @@
+-- 'StakeholderId'.
+lookupExpiryEpoch :: StakeholderId -> VssCertData -> Maybe EpochIndex
+lookupExpiryEpoch id mp = vcExpiryEpoch <$> lookup id mp
+
+-- | Delete certificate corresponding to the specified 'StakeholderId'.
+-- This function is dangerous, because after using it you can't rollback
+-- deleted certificates. Use carefully.
+delete :: StakeholderId -> VssCertData -> VssCertData
+delete id mp =
+    mp & _certs %~ deleteVss id
+       & _whenInsMap %~ HM.delete id
+       & _whenInsSet %~ case mbIns of
+             Nothing  -> identity
+             Just ins -> S.delete (ins, id)
+       & _whenExpire %~ case mbExpiry of
+             Nothing     -> identity
+             Just expiry -> S.delete (expiry, id)
+  where
+    (mbIns, mbExpiry) = lookupEoSes id mp
+
+-- | Set last known slot (lks).
+--   1. If new lks is bigger than 'lastKnownEoS' then some expired certificates
+--      will be removed.
+--   2. If new lks is less than 'lastKnownEoS' then some inserted after @nlks@
+--      certificates will be removed (and 'whenExpire') also will be updated.
+setLastKnownEoS :: EpochOrSlot -> VssCertData -> VssCertData
+setLastKnownEoS nlks mp@VssCertData{..}
+    | nlks > lastKnownEoS = setBiggerLKS nlks mp
+    | nlks < lastKnownEoS = setSmallerLKS nlks mp
+    | otherwise           = mp
+
+setLastKnownSlot :: SlotId -> VssCertData -> VssCertData
+setLastKnownSlot = setLastKnownEoS . EpochOrSlot . Right
+
+-- | Ids of stakeholders issued certificates.
+keys :: VssCertData -> [StakeholderId]
+keys VssCertData{..} = HM.keys (getVssCertificatesMap certs)
+
+-- | Filtering the certificates.
+-- This function is dangerous, because after you using it you can't rollback
+-- deleted certificates. Use carefully.
+filter :: (StakeholderId -> Bool) -> VssCertData -> VssCertData
+filter predicate vcd =
+    foldl' (flip delete) vcd $ List.filter (not . predicate) $ keys vcd
+
+-- | Return True if the specified address hash is present in the map, False otherwise.
+member :: StakeholderId -> VssCertData -> Bool
+member id VssCertData{..} = memberVss id certs
+
+-- | This function is dangerous, because after you using it you can't rollback
+-- deleted certificates. Use carefully.
+difference :: VssCertData -> HM.HashMap StakeholderId a -> VssCertData
+difference mp hm = foldl' (flip delete) mp . HM.keys $ hm
+
+fromList :: [VssCertificate] -> VssCertData
+fromList = foldr' insert empty
+
+----------------------------------------------------------------------------
+-- Helpers
+----------------------------------------------------------------------------
+
+-- | Helper for insert.
+-- Expiry epoch will be converted to expiry slot.
+addInt :: VssCertificate -> VssCertData -> VssCertData
+addInt cert vcd =
+    insertRaw $ expireById False id (addEpoch $ lastKnownEoS vcd) vcd
+  where
+    id = getCertId cert
+    insertRaw mp@VssCertData{..} =
+        let (certs', delCerts) = insertVss cert certs
+        in foldl' (flip delete) mp delCerts
```

Comment:
`foldl' (flip delete)` shows up ~8 times in this PR; does it deserve a name?

Reply:
Not really touching existing code, just moving it from place to place.